Workflows to deploy tagged releases to preprod and prod

lottspot opened the pull request:
These additional workflow jobs implement the following automated actions:

Whenever a release tag is created (runs the 'deploy-preprod' job in workflows/dockerpublish.yml):
- update the k8s/app.yaml manifest with the new version
- deploy the new version to the preprod cluster
- push the updated k8s/app.yaml manifest back up to the develop branch

Whenever a merge happens into releases/v1 ('deploy-prod' job in workflows/deploy.yml):
- deploy the merged k8s/app.yaml manifest to the prod cluster

Working examples:
  - [deploy-preprod](https://github.com/lottspot/chime/runs/534744325?check_suite_focus=true)
  - [deploy-prod](https://github.com/lottspot/chime/runs/534758370?check_suite_focus=true)

Both example jobs were run against the preprod cluster for testing purposes

lottspot committed:
Workflows to deploy tagged releases to preprod and prod

lottspot commented:
It's worth noting that I've already created the necessary secrets for these jobs to work on this repository, so no additional configuration should be required after merge

lottspot commented:
Resolves #48

themightychris commented:
Hmm, in our current [release process](https://codeforphilly.github.io/chime/operations/release-process.html) the merge into `releases/v1` happens first, and then we create the release tag on `releases/v1`, and then we deploy the tag

I think it absolutely make sense for tagged releases to be what gets shipped to preprod, so that it can be an immutable version number that gets reviewed

It also is helpful for a pull request into `releases/v1` to be the approval gate for shipping a release to prod, but then the release tags won't be on the `releases/v1` branch. Maybe that's fine? Seems a bit odd to me though. Would the release tags be on develop then? Do we need to add another branch?

What git flow are you picturing alongside this?

lottspot commented:
Here is my thought process around implementing the workflow that I did:

- The docker build is only happening when there is a tagged release. We cannot deploy to either cluster if there is not a built docker image.
- We know we want prod deploys to happen on merge into the releases/v1 branch. In order for this to happen properly, the develop branch must already have the to-be-deployed version added to k8s/app.yaml when the PR is opened, and it would be ideal if this could automatically track development progress (not requiring ops human intervention)

So with that in mind, I would see the flow looking something like this:

1. Create release tag on a commit in the "develop" branch. This effectively creates a "release candidate". It's built and deployed into preprod.
2. Once release candidate is approved, open PR from develop into releases/v1. On merge, the k8s/app.yaml is applied, which will always have the most recent release candidate in it. So it doesn't actually matter if develop has gotten ahead; it will only actually matter what version is specified in k8s/app.yaml.

It's worth noting that the release process wasn't followed at all for the v1.0.0 release. A tag was simply created and we were asked to deploy it. This gives me the impression that this might be an easier flow for the development team to follow.

themightychris commented:
So if I'm thinking about this right:

1. Team tags a release on develop
2. Docker image gets built and pushed with version tag
3. Manifest gets updated with new version tag
4. Manifest with new version tag gets deployed to preprod
5. A new commit immediately following the tagged one gets pushed to develop, bumping the version in the manifest to the tag version
6. The second commit after the tag gets PRd into `releases/v1` for deploy

It feels a bit odd that we're not deploying the tag, but we kind of are via the container build. I guess that's just an unavoidable consequence of having the infra and source in the same repo, and I can't find anything concretely wrong with it. Merges into develop happening in the middle could be a problem, but the release review would catch that and other flows rely on humans not merging during release anyway

themightychris commented on .github/workflows/dockerpublish.yml:
```diff
@@ -66,3 +69,52 @@ jobs:
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          push-branch: ${{ env.DEPLOY_PUSH_BRANCH }}
+          commit-message: Deploy version ${VERSION} to preprod
```
Let's not mention preprod in the commit message, since the same commit is used to ship to all environments. Maybe just `chore: bump k8s manifest up CHIME ${VERSION}`

lottspot replied:
Good point. Addressed this in 8a553ea

themightychris commented:
Oh also, this doesn't seem to account for us having a sep k8s tree for preprod. I don't think we want to have the whole k8s tree duplicated more than temporarily, but we do need some way for the preprod deploy to use Ingresses with different hostnames so public users can access the preprod deploy for testing

lottspot commented:
> So if I'm thinking about this right:

You've got it right

> I guess that's just an unavoidable consequence of having the infra and source in the same repo

Yeah I think if we are going to deploy the tag there is a lot more we need to change in the workflow. We really need builds deployed to preprod before being deployed to prod, so I think we would need to move toward something more like building an image for every commit on develop, or having a separate release candidate tag which we build for preprod deploys.

> Merges into develop happening in the middle could be a problem

Thought about this, but it's not likely that we run into problems with this since the action is only making one line changes to k8s/app.yaml, and we're to the point where humans are basically not touching that file, minimizing our chances for problems in practice. Worst case scenario, there is a conflict which halts the deploy process and a human has to resolve it.

> release review would catch that

Techinically, no matter how many changes are merged into develop between the time a release candidate is tested in preprod and merged into releases/v1, those changes will not reach production, since the deploy job will only apply whatever is already in the k8s/app.yaml file, and that file is only updated whenever a release candidate would be tagged.

> we do need some way for the preprod deploy to use Ingresses with different hostnames

This is not a new problem introduced by this flow. This is an existing problem regardless of whether this automation is implemented or not. There are two separate jobs for deploying to preprod vs deploying to prod, so we can have unique steps for each process, e.g., applying different ingress resources within each job.

lottspot committed:
Version bumps in k8s/app.yaml are not environment specific

lottspot commented:
>  no matter how many changes are merged into develop between the time a release candidate is tested in preprod and merged into releases/v1, those changes will not reach production

This is of course not to say that we couldn't do better here. It is obviously awkward to have code living in releases/v1 that is not actually part of the release. One possibility is to create a release candidate branch off of the tag (or maybe just have each release candidate be a branch instead of a tag?) and push the manifest update back to the rc branch, and then merge the rc branch into releases/v1. That would probably address a lot of the inherent awkwardness in the proposed flow, but we would need buy in from the app team.

themightychris commented:
@lottspot I'm not saying we need to change anything at this point and this seems like a good starting point at least. Just thinking through potential implications out loud to make sure we're not missing any gotchas

themightychris commented:
> This is not a new problem introduced by this flow. This is an existing problem regardless of whether this automation is implemented or not. 

Well there is a pretty brute-force solution merged in (#219) maybe since you started this branch of copying the entire `k8s/` tree to `k8s-preprod/`. The `preprod` variant is what's deployed right now and working, so merging this as-is _would_ break preprod I believe.

I want to find a better solution to that whole tree being duplicated sooner than later, two quick fixes we could potentially do within this PR after you rebase it:

1. As [I noted in the preprod PR after merging it](https://github.com/CodeForPhilly/chime/pull/219#issuecomment-603986304) we could break out the ingress definitions into their own file and symlink everything else in `k8s-preprod/*` to `k8s/*`
2. Or to avoid using symlinks, we could do the same ingress split-out and break up `k8s/` into `k8s/`, `k8s.prod/`, and `k8s.prepod/` and then make all our deploys be a two-step `kubectl apply -Rf k8s/ && kubectl apply -Rf k8s.${env}/`

lottspot committed:
Minimize difference between prod and preprod manifest trees

lottspot committed:
Remove duplicated preprod tree

lottspot commented:
> so merging this as-is would break preprod I believe

Ah, but little did you know, I already broke the hell out of preprod by brutishly deploying the prod manifests when I was testing these workflows 😄 

> Or to avoid using symlinks, we could do the same ingress split-out and break up k8s/ into k8s/, k8s.prod/, and k8s.prepod/

I like this quick fix. I implemented it in 63bf8c1 and removed the duplicate preprod tree in 12ba277

lottspot committed:
Fix version reference in git commit job

lottspot commented:
Updated versions of these jobs have been tested:

- [deploy-preprod](https://github.com/lottspot/chime/runs/536626529?check_suite_focus=true)
- [deploy-prod](https://github.com/lottspot/chime/runs/536642430?check_suite_focus=true) ([PR](https://github.com/lottspot/chime/pull/13))

lottspot committed:
Show kubernetes resource diffs before merging release

lottspot commented:
7be88df adds a show-diff job which will publish the output of `kubectl diff` to a comment on any PR opened into releases/v1. A successful test run of this job can be seen at: https://github.com/lottspot/chime/pull/15#issuecomment-605697110

themightychris commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,107 @@
+          To be applied to prod from $app_manifest:
+          \`\`\`
+          $app_diff
+          \`\`\`
```
```suggestion
          \`\`\`diff
          $app_diff
          \`\`\`
```

themightychris replied:
this works! see https://github.com/lottspot/chime/pull/15#issuecomment-605704465

themightychris commented on .github/workflows/deploy.yml:
````diff
@@ -0,0 +1,107 @@
+          To be applied to prod from $prod_manifest_dir:
+          ```
+          $prod_diff
+          ```
````
```suggestion
          ```diff
          $prod_diff
          ```
```

lottspot committed:
Color code prod_diff in deploy workflow

Co-Authored-By: Chris Alfano <calfano@codeforamerica.org>

lottspot committed:
Color code app_diff in deploy workflow

Co-Authored-By: Chris Alfano <calfano@codeforamerica.org>

MooseQuest reviewed (commented):
Could I get confirmation that you move the file structure to separate file folders. Just want to confirm that. @lottspot

lottspot commented:
@MooseQuest yes the file structure has been split into three different directories now:

- `k8s`: Common manifests to apply to both preprod and prod
- `k8s.preprod`: Manifests which should only be applied to the preprod cluster
- `k8s.prod`: Manifests which should only be applied to the prod cluster

lottspot commented:
Agreed to give this merge a go via slack convo: https://codeforphilly.slack.com/archives/C0107KSP9QF/p1585580958213600
